src/python/pants/backend/docgen/tasks/markdown_to_html_utils.py

```diff
@@ -66,57 +72,91 @@ def choose_include_text(s, params, source_path):
   :param params: string like "start-at=foo&end-at=bar"
   :param source_path: path to source .md. Useful in error messages
   """
-  lines = s.splitlines()
-  start_after = None
-  start_at = None
-  end_before = None
-  end_at = None
+  source_lines = s.splitlines()
+  params_dict = {}
 
   for term in params.split("&"):
     if '=' in term:
       param, value = [p.strip() for p in term.split('=', 1)]
     else:
       param, value = term.strip(), ''
     if not param: continue
-    if param == "start-after":
-      start_after = value
-    elif param == "start-at":
-      start_at = value
-    elif param == "end-before":
-      end_before = value
-    elif param == "end-at":
-      end_at = value
+    if param not in INCLUDE_RECOGNIZED_PARAMS:
+      raise TaskError('Invalid include directive "{0}" in params "{1}"'
+                      ' from {2}'.format(param, params, source_path))
+    params_dict[param] = value
+
+  chosen_lines = []
+  try:
+    if 'lines' in params_dict:
+      lines_arg = params_dict['lines']
+      chosen_lines = include_by_line_range(source_lines, lines_arg, params_dict)
```

Comment:
There is a minor asymmetry to `include_by_line_range` and `include_by_text_match` in that one takes pre-parsed arguments from the params_dict, and the other doesn't. I'd recommend pre-parsing the entire params_dict in this function, and then passing the parsed params into those functions rather than the params_dict.

Reply:
I like this a lot. It also sidesteps the concerns I had below about handling errors in `include_by_*` (because error handling can be kept mostly in `choose_include_text`), so I think I'll do this instead.